runtime/gc_vlhgc/CopyForwardScheme.cpp

```diff
@@ -3269,31 +3323,167 @@ MM_CopyForwardScheme::completeScanForAbort(MM_EnvironmentVLHGC *env)
 	} while (handleOverflow(env));
 }
 
+void
+MM_CopyForwardScheme::completeScanWorkPacket(MM_EnvironmentVLHGC *env)
+{
+	MM_AllocationContextTarok *reservingContext = _commonContext;
+	J9Object *objectPtr = NULL;
+
+	while (NULL != (objectPtr = (J9Object *)env->_workStack.popNoWaitFromCurrentInputPacket(env))) {
+		Assert_MM_false(MM_ScavengerForwardedHeader(objectPtr).isForwardedPointer());
+		scanObject(env, reservingContext, objectPtr, SCAN_REASON_PACKET);
+	}
+}
+
+MM_CopyForwardScheme::CopyForwardSchemeScanReason
+MM_CopyForwardScheme::getNextWorkUnit(MM_EnvironmentVLHGC *env, UDATA preferredNumaNode)
+{
+	env->_scanCache = NULL;
+	CopyForwardSchemeScanReason ret = SCAN_REASON_NONE;
+
+	MM_CopyScanCacheVLHGC *cache = NULL;
+	/* Preference is to use survivor copy cache */
+	if(NULL != (cache = getSurvivorCacheForScan(env))) {
+		env->_scanCache = cache;
+		ret = SCAN_REASON_COPYSCANCACHE;
+		return ret;
+	}
+
+	if (NULL != env->_deferredScanCache) {
+		/* there is deferred scanning to do from partial depth first scanning */
+		cache = (MM_CopyScanCacheVLHGC *)env->_deferredScanCache;
+		env->_deferredScanCache = NULL;
+		env->_scanCache = cache;
+		ret = SCAN_REASON_COPYSCANCACHE;
+		return ret;
+	}
+
+#if defined(J9MODRON_TGC_PARALLEL_STATISTICS)
+	env->_copyForwardStats._acquireScanListCount += 1;
+#endif /* J9MODRON_TGC_PARALLEL_STATISTICS */
+
+	UDATA nodeLists = _scanCacheListSize;
+	bool doneFlag = false;
+	volatile UDATA doneIndex = _doneIndex;
+
+	while ((SCAN_REASON_NONE == ret) && !doneFlag) {
+		/* local node first */
+		ret = getNextWorkUnitOnNode(env, preferredNumaNode);
+		if (SCAN_REASON_NONE == ret) {
+			/* we failed to find a scan cache on our preferred node */
+			if (COMMON_CONTEXT_INDEX != preferredNumaNode) {
+				/* try the common node */
+				ret = getNextWorkUnitOnNode(env, COMMON_CONTEXT_INDEX);
+			}
+			/* now try the remaining nodes */
+			UDATA nextNode = (preferredNumaNode + 1) % nodeLists;
+			while ((SCAN_REASON_NONE == ret) && (nextNode != preferredNumaNode)) {
+				if (COMMON_CONTEXT_INDEX != nextNode) {
+					ret = getNextWorkUnitOnNode(env, nextNode);
+				}
+				nextNode = (nextNode + 1) % nodeLists;
+			}
+		}
+		if (SCAN_REASON_NONE == ret && (0 != _regionCountCannotBeEvacuated) && !_abortInProgress && !abortFlagRaised()) {
+			if (env->_workStack.retrieveInputPacket(env)) {
+				ret = SCAN_REASON_PACKET;
+			}
+		}
```

Comment:
for code cleanness, lines from 3371 to 3391 should go into method called something like getNextWorkUnitNoWait(), that does not take any NUMA node param.

for now, the order we consume things  is ok (prefer copy queue, because realistically more regions are to be evacuated), but in future changes, we have to be fair to both copy and mark work Q. we should somehow interleave. something like: pick from whichever is currently more likely to overflow.

Reply:
done